### PR TITLE
Type-stabilizing array concatenations

### LINKED ISSUE #13665: Inference failure for concatenation
``` julia
function test()
    x=[1. 2.]
    y=3
    result = [y x]
    return result
end
@code_warntype test()
```

results in

```
Variables:
  x::Array{Float64,2}
  y::Int64
  result::ANY
  ##X#9661::Tuple{Int64,Array{Float64,2}}
  ####X#7774#9662::Tuple{Int64,Array{Float64,2}}
  ####T#7775#9663::TYPE{T}

Body:
  begin  # In[150], line 2:
      x = (Main.hcat)(1.0,2.0)::Array{Float64,2} # In[150], line 3:
      y = 3 # In[150], line 4:
      GenSym(1) = y::Int64
      GenSym(0) = x::Array{Float64,2}
      GenSym(2) = AST(:($(Expr(:lambda, Any[:(x::Any::ANY)], Any[Any[Any[:x,Any,0]],Any[],0,Any[]], :(begin  # abstractarray.jl, line 798:
        unless (Base.isa)(x,Base.AbstractArray)::ANY goto 0
        return (Base.eltype)(x)::ANY
        0: 
        return (Base.typeof)(x)::ANY
    end::ANY)))))
      ####T#7775#9663 = (Base.promote_type)((GenSym(2))(GenSym(1))::ANY,(GenSym(2))(GenSym(0))::ANY)::TYPE{T}
      result = (Base.cat_t)(2,####T#7775#9663::TYPE{T},GenSym(1),GenSym(0))::ANY # In[150], line 5:
      return result
  end::ANY
```

This issue looks related to, but different from #13254 since there is no `T[a:b;]` involved.

See also discussion at https://groups.google.com/forum/#!topic/julia-users/_lIVpV0e_WI

### LINKED ISSUE #19038: Type stability of array concatenations
Both

```
@code_warntype [[1.0, 2.0]; 3.0]
```

and

```
@code_warntype vcat([1.0, 2.0], 3.0)
```

give `Any` as final inferred type.

Is this expected? Or must this be an issue? I couldn't find any previous issue on this...

### PR DESCRIPTION
**Pull request by p-zubieta**
This is a rewriting of the some of the array concatenation methods to make them type stable. The PR adds a type stable version of `cat` (`cat{n}(::Type{Val{n}}, X...)`) while preserving the existing API (which cannot be made inferable, but should be just a bit faster than before anyway).

Fixes #13665,  #19038 and #19304

NOTE: I believe the first three commits could be backported to 0.5.

### CODE REVIEW

File: base/abstractarray.jl [hunk 1]
```diff
@@ -1095,55 +1095,76 @@ end
...
+
+function cat_t(dims, T::Type, X...)
+    sifter = dims2sift(dims)
+    shape = cat_shape(sifter, length(X), (), map(cat_size, X)...)
+    return _cat(T, shape, sifter, X...)
+end
+
+function _cat(T::Type, shape, sifter, X...)
+    N = length(shape)
+    A = cat_similar(X[1], T, shape)
```
Comment (nalimilan) [inline base/abstractarray.jl]: While you're at it, could you choose the return type based on all input types (https://github.com/JuliaLang/julia/issues/2326)? Or do you think it should go into another PR?

Comment (pabloferz) [inline base/abstractarray.jl]: I'd like to handle that somehow, but I'd say that it would have to be addressed elsewhere. For one I have some ideas that would require the new sub-typing algorithm in place.

**Comment by stevengj** [comment]
Do we have benchmark coverage of these functions for dense and sparse matrices?

**Comment by stevengj** [comment]
Needs tests?

File: base/sparse/sparsevector.jl [hunk 2]
```diff
@@ -872,16 +872,16 @@ typealias _TypedDenseConcatGroup{T} Union{Vector{T}, Matrix{T}, _Annotated_Typed
...
     Base.cat_t(catdims, T, X...)
 end
 function hcat(Xin::_SparseConcatGroup...)
-    X = SparseMatrixCSC[issparse(x) ? x : sparse(x) for x in Xin]
+    X = map(x -> SparseMatrixCSC(issparse(x) ? x : sparse(x)), Xin)
     hcat(X...)
 end
 function vcat(Xin::_SparseConcatGroup...)
-    X = SparseMatrixCSC[issparse(x) ? x : sparse(x) for x in Xin]
+    X = map(x -> SparseMatrixCSC(issparse(x) ? x : sparse(x)), Xin)
```
Comment (Sacha0) [inline base/sparse/sparsevector.jl]: Out of curiosity, why replace the comprehensions with `map`s? Generate `Tuple`s rather than `Vector`s?

Would `SparseMatrixCSC` suffice in place of `x -> SparseMatrixCSC(issparse(x) ? x : sparse(x))`?

Best!

Comment (pabloferz) [inline base/sparse/sparsevector.jl]: > Out of curiosity, why replace the comprehensions with maps? Generate Tuples rather than Vectors?

Just to avoid the allocation o creating the array, although it shouldn't make much of a difference.

> Would SparseMatrixCSC suffice in place of x -> SparseMatrixCSC(issparse(x) ? x : sparse(x))?

`SparseMatrixCSC([1])` is not the same as `SparseMatrixCSC(sparse([1]))`. Actually, the first one does not work.

**Comment by pabloferz** [comment]
I don't there are benchmarks for concatenations of mixed dense and sparse matrices. Is it worth to check the performance anyway?

File: base/sparse/sparsevector.jl (hunk 2 above)
Comment (nalimilan) [inline base/sparse/sparsevector.jl]: `map` returns a tuple for tuple inputs, while a comprehension creates an array.

Comment (Sacha0) [inline base/sparse/sparsevector.jl]: Your modifications look great to me --- thank you for the explanations!

**Comment by pabloferz** [comment]
I think we can remove the 'needs tests' label, there already tests covering the pointed issues.

**Comment by pabloferz** [comment]
Unless there is an impact on performance or there are any more comments, this is ready on my side.

**Comment by stevengj** [comment]
It needs `@inferred` tests for the type stability, to prevent regressions.

**Comment by pabloferz** [comment]
There are already tests in there, but I can add more if they don't seem enough.

**Comment by stevengj** [comment]
@pabloferz, I just want to make sure that tests were added for any issues that you fixed, e.g. I don't see a test for #19304.

File: base/abstractarray.jl [hunk 3]
```diff
@@ -1095,55 +1095,76 @@ end
...
+dims2cat{n}(::Type{Val{n}}) = ntuple(i -> (i == n), Val{n})
+dims2cat(dims) = ntuple(i -> (i in dims), maximum(dims))
+
+cat(dims, X...) = cat_t(dims, promote_eltype(X...), X...)
+
+function cat_t(dims, T::Type, X...)
+    catdims = dims2cat(dims)
+    shape = cat_shape(catdims, (), map(cat_size, X)...)
+    A = cat_similar(X[1], T, shape)
+    if countnz(catdims) > 1 && T <: Number
```
Comment (pabloferz) [inline base/abstractarray.jl]: Good call. Changed.

**Comment by stevengj** [comment]
LGTM.

**Comment by tkelman** [comment]
Why wasn't nanosoldier run here?

**Comment by stevengj** [comment]
Forgot, sorry.

**Comment by pabloferz** [comment]
Fortunately, there were only speed improvements that came from this. See https://github.com/JuliaCI/BaseBenchmarkReports/blob/a9c6ffef26b60d527d06b80ac3ea2fde79637a2a/daily_2016_12_2/report.md

**Comment by tkelman** [comment]
This broke quite a few packages: https://htmlpreview.github.io/?https://github.com/JuliaCI/pkg.julialang.org/blob/ac017050e0c662e46feef496e77ac12baa85583c/pulse.html

I haven't bisected all of them, so this isn't necessarily at fault for every one, but it's at least the underlying cause for AverageShiftedHistograms and BlackBoxOptim.

**Comment by nalimilan** [comment]
The new failure in CategoricalArrays tests is due to the fact that `vcat(["a"], "az")` now returns a `Vector{Any}`, when it previously returned a `Vector{String}` (both are `==`). Of course this can (should?) be written as `vcat(["a"], ["az"])`, but I'm not sure whether `Any` is intended or not. For example, `vcat([1], 2)` still returns a `Vector{Int}`.

**Comment by pabloferz** [comment]
@nalimilan That's because we this uses now `promote_eltype` for type stability reasons, but `eltype(String) == Char`. So we need special handling for scalars (in this case non `AbstractArray` subtypes before doing the `promote_eltype`). Fortunately, that is pretty easy to fix.

@tkelman I'll look into the ones you mentioned above to see what is the problem, and I you get a bigger list I can look into that too.

**Comment by pabloferz** [comment]
@tkelman I didn't check them all, but #19523 seems to fix the problems on a bunch of packages.
